### ISSUE TITLE
implement mesh instancing

### ISSUE THREAD

**Issue by Ziriax** [Ziriax/Maya2glTF#10]
Currently each instanced maya shape is exported as a new mesh. Most code should be in place to deal with instances.

**Comment by Ziriax** [comment]
I have some test scenes for instancing in this repository

I will first test these again to check if these still work.

**Comment by Ziriax** [comment]
Indeed, mesh instancing doesn't seem to work. But it should work, so this is a bug.

**Comment by Ziriax** [comment]
Ah, no,  this is not yet implemented, the comment says why

```
    // TODO: Support instancing, for now we create a new mesh for each node.
    // To properly support instance, we need to decide what to do with shapes
    // that are both with and without a skeleton Do we generate two meshes, with
    // and without skinning vertex attributes?
    ExportableMesh(ExportableScene &scene, ExportableNode &node,
                   const MDagPath &shapeDagPath);
```

**Comment by Ziriax** [comment]
Bummer, I forgot about this...

This is a very important feature, and should be implemented.

That being said, do you think this plugin is still relevant? The BabylonJS team is much larger and has made their own plugin...

**Comment by Ziriax** [comment]
Maybe the easiest approach is hashing the vertex element buffers, and what buffers a mesh uses. If all buffers are the same, then instance it. This solves the issue with weights etc... The mesh name could also be considered to avoid different Maya meshes to become instances. 

As a first step, the element buffers should always be shared if possible, since these are read-only by definition I guess, so it doesn't make sense to store these twice...

**Comment by BigRoy** [comment]
> Maybe the easiest approach is hashing the vertex element buffers, and what buffers a mesh uses. If all buffers are the same, then instance it. This solves the issue with weights etc... The mesh name could also be considered to avoid different Maya meshes to become instances.
> 
> As a first step, the element buffers should always be shared if possible, since these are read-only by definition I guess, so it doesn't make sense to store these twice...

Actually no. The problem is much more trivial. You can just query Maya if the shape is instanced and if it is you can also find the other instances (and e.g. the master instance - which Maya doesn't really have; instancing is just the same mesh shape parented under multiple transforms).

Thus, with instancing in Maya you can't have one mesh being skinned and the other not being skinned if they were supposed to actual instances - because any case where the shape node differs is basically not an instance.

Regarding BabylonJS. I haven't used the exporter yet but only read through the docs. Because it first goes through the Babylon engine for conversion it felt bloated and limited to what the engine supports at best. Could be decent enough for most use cases however. Hard to tell whether it's worth dropping this particular plugin. If this one isn't being maintained I can see how the other could outrun it.

I was thinking of maybe rewriting this full exporter with the Maya python API just because I feel Maya users are more familiar in that area and it being Python it might get more development support. (But I just don't use it enough for work to warrant spending so much time getting that involved.)

**Comment by Ziriax** [comment]
That could work, but Maya also supports transform instances, not just shapes, something GLTF doesn't support. So basically any transform can have any number of parents in Maya, that's how they do instances. So extra logic will be needed I guess. But possible

The problem is that it is possible in Maya (as far as I remember) to have a mesh that is skinned, and other instance that isn't, and in GLTF, that can't be the same mesh, so instancing can't work in that case. But the accessors can be the same, and hence on the GPU, the same buffers could be re-used. 

No this plugin isn't really maintained, but I try to help out people with requests, and like to work on it in my free time. But it is not actively used by the original company/sponsor anymore (they switched to Unity3D and FBX). Personally I'm unfortunately not doing any 3D coding anymore.

Rewriting in Python might be a good idea, but I won't be able to help with that, I just hate Python, any untyped language for that matter 😉
